Avoid NuGet restore cycles

ocallesp opened the pull request:
Fixes #8326

Avoid infinite NuGet restores.
Scenario:
Projects might have a package that sets `AssetTargetsFallback` value that can cause a certain `PackageReference` to be included. This package contains a props file that changed the `AssetTargetFallback` property value, and the `PackageReference` Condition was no longer true, causing it to be removed. Once removed, the `AssetTargetFallback` value from the `csproj` was used again, causing an infinite restore -> design time build -> restore loop.

This change:
Since DotNet ProjectSystem can detect changes in NuGet restores by calculating a hash, we can use this hash to detect changes in `AssetTargetFallBack` and check when this value gets repeated.

This pr makes use of the `NuGetRestoreCycleDetector` to keep the last N hashes, and once it detects a repeated value (cycle) Dotnet ProjectSystem can stop NuGet restores.


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/project-system/pull/8239)

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            lock (_lock)
+            {
+                foreach (var b in _values)
+                {
+                    if (AreEqual(b, hashValue))
+                    {
+                        return true;
+                    }
+                }
```
This loop detects whether the current hash has been seen within the last N hashes.

You could trigger this by adding a package reference, then removing it. Once removed, the project is back in the prior state and restore should complete.

With this change, I believe we would not restore after removing the recently added package reference.

As an alternative, consider something like:

- store the last N hashes
- each time a new hash is produced
  - append it to the list
  - discard any hashes added longer than some threshold ago
  - walk back through the list, looking for cycles of repeating patterns of length 2, 3, 4, ...
  - if a repeating pattern is found, there is a cycle

I'm not 100% sure about that approach, but I think it would be more robust than the approach here.

nkolev92 replied:
> I am wondering if NuGet can easily detect a bad package

What would be a bad package in this case?

If you are referring to a package with props/targets that change the inputs, the MSBuild side is hidden from NuGet, so NuGet has no means of figuring it out, so this logic is technically handling that.

drewnoakes replied:
> I modified the logic a little to detect the cycle only when counter == Size.

I'm concerned that this is only matching values and not looking for patterns.

For example (where each letter represents a hash):

- A B A B A B A B A B A B
- A B C A B C A B C A B C
- A B A C A D E A F A G A

The first two sequences look more like infinite cycles to me than the third one, yet the third would trip the check here.

ocallesp replied:
@nkolev92 do you think these patterns are possible to appear ?

nkolev92 replied:
In theory, you can come up with packages/authoring that has a pattern of any length. 

In my experience, I've only seen  A B A B patterns.  

The most common occurrences of continuous restores is the dynamically generate version one. So maybe we try to detect situations where say a bunch of consecutive design time builds lead to a restore nomination. If we want to special case that, we can probably use the NuGet start/stop restore markers for help.

ocallesp replied:
@nkolev92  if we are in a loop, does it mean that we already completed NuGet restore ?

nkolev92 replied:
yep. 

The loop is effectively, design time build which nominates, restore, design time build which nominates, restore.

ocallesp replied:
@nkolev92 at the project system level we don't know what package is causing the loop. We can show a yellow bar to notify the user that one of the nuget packages is invalid. Is there a way to identify the package from the NuGet side ? so we can get the name of the package from NuGet and use it in the yellow bar ?

nkolev92 replied:
> @nkolev92 at the project system level we don't know what package is causing the loop. We can show a yellow bar to notify the user that one of the nuget packages is invalid and is causing a nuget restore cycle.

Yep, that's in line with what I'd expect based on how MSBuild works :) 

>  Is there a way to identify the package from the NuGet side ? so we can get the name of the package from NuGet and use it in the yellow bar ?

NuGet doesn't really anything MSBuild in VS, so there's pretty much no way of figuring it out.

ocallesp replied:
Done.
An InfoBar will be displayed in the main window with the message "NuGet restore detected a cycle. NuGet Error NU1108"

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -134,6 +135,10 @@ private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
                 _latestHash = hash;
+                if (_cycleDetector.ComputeCycleDetection(hash))
+                {
+                    return true;
```
We should surface the fact that a loop is suspected to the user, probably via a gold bar.

Something like "An infinite NuGet restore loop is suspected in project _MyProject_. This can happen when a package changes the project's imports, which requires a further restore. The most recent restore was skipped in order to break the cycle."

This serves two purposes:

1. Highlights the issue to the user in case they wish to fix it. It would be great if we could identify the package causing the issue to further help the user.
2. Shows the user that a restore was cancelled, which may be important if they do not actually have a cycle. Perhaps they were just switching between two branches several times, for example.

drewnoakes replied:
> Shows the user that a restore was cancelled, which may be important if they do not actually have a cycle. Perhaps they were just switching between two branches several times, for example.

Perhaps we add a button to the infobar to allow forcing a restore. Similar to how CPS's project fault infobar has a button to reload the project.

ocallesp replied:
InfoBar added.

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -49,6 +49,7 @@ internal partial class PackageRestoreDataSource : ChainedProjectValueDataSourceB
         // |_________________________________________|    |_________________________________________|
         //
 
+        private readonly NuGetRestoreCycleDetector _cycleDetector = new(5);
```
Do we have a sense of how long it would take to restore five times? If it's not very long, I'd vote to increase this threshold.

nkolev92 replied:
You can maybe use https://docs.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio#ivsnugetprojectupdateevents-interface to detect the timing of restore start/end. 

We can also add other events if they'd make life easier for you to do detect these loops.

ocallesp replied:
In the next commit the threshold will be increased to 10

drewnoakes commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs:
```diff
@@ -0,0 +1,90 @@
+                }
+                else
+                {
+                    Add(hashValue);
```
If the queue does not contain this value, I think the counter should be reset to zero.

Also the length of the queue should probably be capped. Otherwise the probability of finding the hash in the queue increases over time.

ocallesp replied:
done.

nkolev92 commented:
Any updates on this change @ocallesp?

MiYanni commented:
@ocalles You're getting build failures with these changes.
```
(CoreCompile target) -> 
         D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(7,44): error CS0234: The type or namespace name 'Implementation' does not exist in the namespace 'Microsoft.VisualStudio.ProjectSystem' (are you missing an assembly reference?)
         D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(55,26): error CS0246: The type or namespace name 'IInfoBarService' could not be found (are you missing a using directive or an assembly reference?)
         D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(68,13): error CS0246: The type or namespace name 'IInfoBarService' could not be found (are you missing a using directive or an assembly reference?)
```

ocallesp commented:
> @ocalles You're getting build failures with these changes.
> 
> ```
> (CoreCompile target) -> 
>          D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(7,44): error CS0234: The type or namespace name 'Implementation' does not exist in the namespace 'Microsoft.VisualStudio.ProjectSystem' (are you missing an assembly reference?)
>          D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(55,26): error CS0246: The type or namespace name 'IInfoBarService' could not be found (are you missing a using directive or an assembly reference?)
>          D:\a\_work\1\s\src\Microsoft.VisualStudio.ProjectSystem.Managed.VS\ProjectSystem\VS\PackageRestore\PackageRestoreDataSource.cs(68,13): error CS0246: The type or namespace name 'IInfoBarService' could not be found (are you missing a using directive or an assembly reference?)
> ```

Fixed.
I added the implementation of the InfoBar service as part of this pr.

ocallesp commented:
Tests:
Open a solution having an invalid NuGet package. Cycle was detected.
Open roslyn and orchardcore. No cycles detected.

ocallesp commented:
> Any updates on this change @ocallesp?

Today I add a new commit with the implementation of InfoBar. This pr is complete. Waiting for reviewers and approval

kvenkatrajan commented:
@ocallesp - not required for the initial merge but anyways we can add telemetry around this for when we detect a cyclic restore? Will help determine how often our users run into this scenario, we detect it & display the infobar

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/Telemetry/TelemetryEventName.cs:
```diff
@@ -71,5 +71,10 @@ internal static class TelemetryEventName
+        /// <summary>
+        ///     Indicates that the NuGet restore detected a cycle.
+        /// </summary>
+        public const string NuGetRestoreCycleDetected = Prefix + "/nugetrestore/cycledetected";
```
>cycledetected [](http://example.com/codeflow?start=80&length=13)

Don't forget to get this blessed once it starts coming through.

kvenkatrajan replied:
think it would be helpful to add additional properties in addition to cycledetected. So have the event name as : 
/nugetrestore/cycledetection and properties like vs.projectsystem.managed.nugetrestore.cycledetection.cycledetected, vs.projectsystem.managed.nugetrestore.cycledetection.durationmillis, vs.projectsystem.managed.nugetrestore.cycledetection.successfulrestorecount,
vs.projectsystem.managed.nugetrestore.cycledetection.projectid or vs.projectsystem.managed.nugetrestore.cycledetection.solutionid perhaps. Any additional items you think may be useful to help troubleshoot when we detect a cyclical restore.

jacdavis replied:
I *believe* project id and solutionid are sent automatically. Duration and succesffulrestorecount would be nice. I think succesffulrestorecount  will be unreliable as I believe we must stop as soon as a cycle is detected.

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/VSShellServices.cs:
```diff
@@ -0,0 +1,67 @@
+    {
+        vsShell = ServiceProvider.Value;
+
+        if (ErrorHandler.Succeeded(vsShell.GetProperty((int)__VSSPROPID.VSSPROPID_IsInCommandLineMode, out object result)) && (bool)result)
```
>if (ErrorHandler.Succeeded(vsShell.GetProperty((int)__VSSPROPID.VSSPROPID_IsInCommandLineMode, out object result)) && (bool)result) [](http://example.com/codeflow?start=8&length=131)

Are we sure this always executes on the ui thread? IIRC, IVsShell  requires the ui thread and doesn't have a proxy.

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/VSShellServices.cs:
```diff
@@ -0,0 +1,67 @@
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private bool CheckIsInServerMode()
+    {
+        if (ErrorHandler.Succeeded(vsShell.GetProperty((int)__VSSPROPID11.VSSPROPID_ShellMode, out object value))
```
>GetProperty [](http://example.com/codeflow?start=43&length=11)

Same as above. What thread does this execute on? I would feel better about a JTF call to ensure we're on the ui thread.

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.InfoBarEntry.cs:
```diff
@@ -0,0 +1,61 @@
+            Assumes.NotNull(items);
+            Assumes.NotNull(onClose);
+
+            Message = message;
```
>Message [](http://example.com/codeflow?start=12&length=7)

normally, C# static analysis will throw a warning if fields are accessed without "this.". ISt his disabled in this codebase?

jacdavis replied:
note that I'm having a hrad time understanding this code without the this. prefixes. For instance, I assumed cookie was a local in this function.

jacdavis replied:
It looks like there is existing code that doesn't put in the "this." for property invocation, but fields all use the _ prefix.

tmeschter replied:
Fields should use the _ prefix.

ocallesp replied:
prefix added. Fixed in last commit

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.cs:
```diff
@@ -0,0 +1,132 @@
+        Requires.NotNullOrEmpty(message, nameof(message));
+        Requires.NotNull(items, nameof(items));
+
+        if (!_vsShellServices.IsInServerMode)
```
>if (!_vsShellServices.IsInServerMode) [](http://example.com/codeflow?start=8&length=37)

What is the right behavior if this is in server mode? Right now, it silently succeeds.

jacdavis replied:
Are you sure IsInServerMode works off the ui thread? This is invoked on a arbitrary thread in cycle detector.

kvenkatrajan replied:
IsInServerMode -> is this to account for Nexus based scenarios? Do we need to explicitly call out a NotSupportedException when IsInServerMode = true?

tmeschter replied:
This was originally to account for Nexus scenarios, yes, and as we don't really have a server SKU under active development it probably doesn't matter much. I think it's reasonable to throw an exception if we are in server mode, and also agree with @jacdavis that we need to verify `IsInServerMode` is ok to call from an arbitrary thread, and/or transition to the UI thread.

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.cs:
```diff
@@ -0,0 +1,132 @@
+
+    private void RemoveInfoBar(string message)
+    {
+        // Assumption is that message is "good" enough to uniquely identify problems
```
> // Assumption is that message is "good" enough to uniquely identify problems [](http://example.com/codeflow?start=7&length=77)

Are we sure this is the case? Seems like a bug waiting to happen

ocallesp replied:
We just want to search if there is an existing InfoBar with the same message (not taking into account buttons/links). We can change this implementation if that is really needed.

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -49,6 +52,9 @@ internal partial class PackageRestoreDataSource : ChainedProjectValueDataSourceB
         // |_________________________________________|    |_________________________________________|
         //
 
+        private readonly NuGetRestoreCycleDetector _cycleDetector = new(10);
```
>10 [](http://example.com/codeflow?start=72&length=2)

What is this magic number? Why 10 here?

ocallesp replied:
5 is enough to detect this scenario
X -> Y -> X -> Y -> X

jacdavis replied:
Please don't use a magic number. Put it in a named constant that explains the reasoning behind it, or at a minimum, add a comment about why this number was chosen.

ocallesp replied:
fixed

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs:
```diff
@@ -0,0 +1,90 @@
+                {
+                    counter++;
+                    
+                    // Verify that a hashValue has repeted in almost all cases
```
>repeted [](http://example.com/codeflow?start=51&length=7)

spelling

ocallesp replied:
fixed

jacdavis commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/NuGetRestoreCycleDetector.cs:
```diff
@@ -0,0 +1,90 @@
+    internal sealed class NuGetRestoreCycleDetector
+    {
+        private readonly object _lock;
+        private readonly Queue<byte[]> _values;
```
>Queue [](http://example.com/codeflow?start=25&length=5)

Why are we using a queue here and not some data structure with O(1) lookup? I expected this to be a hash table of hashes. A queue doesn't make a lot of sense in my mind. Am I missing some info?

ocallesp replied:
I added the hash _lookupTable to improve the search to O(1)

jacdavis replied:
I'm still not convinced a queue is the correct data structure here. Can you please call me to discuss as requested?

nkolev92 commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -135,6 +145,17 @@ private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
 
                 _latestHash = hash;
 
+                if (_cycleDetector.ComputeCycleDetection(hash))
```
Even if we hit this here, still nominate on line 160 right?

That should probably not happen correct?

ocallesp replied:
I think we should return succeed = false.

ocallesp commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -139,6 +139,7 @@ private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
 
                 if (_latestHash != null && hash.AsSpan().SequenceEqual(_latestHash))
                 {
+                    _cycleDetector.Clear();
```
@nkolev92 with this change we avoid showing the infobar in the scenario where
a user adds a nuget package, then delete the nuget package, then add the same package, then delete the same package.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs:
```diff
@@ -2,9 +2,12 @@
 using Microsoft.VisualStudio.Threading;
+using Microsoft.VisualStudio.ProjectSystem.VS.UI.InfoBarService;
 using NuGet.SolutionRestoreManager;
+using Microsoft.VisualStudio.Telemetry;
```
Nit: sort the using statement.

ocallesp replied:
sorted.
Fixed in last commit

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBarService/VsInfoBarService.cs:
```diff
@@ -0,0 +1,132 @@
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.PackageRestore.InfoBarService;
```
The InfoBarService isn't really specific to package restore scenarios; I suggest moving this to a different namespace and folder.

ocallesp replied:
Umm.... I thought this was updated by the NamespaceUpdate.

I'll fix this in the next commit.

ocallesp replied:
fixed in last commit

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/Shell/VSShellServices.cs:
```diff
@@ -0,0 +1,67 @@
+        if (disposing)
+        {
+        }
+    }
```
Is this method dead code?

ocallesp replied:
deleted in last commit.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/VSResources.resx:
```diff
@@ -295,4 +295,7 @@ In order to debug this project, add an executable project to this solution which
     <comment>{0} is the line number of the problematic XAML, and {1} is the column.</comment>
   </data>
+  <data name="InfoBarMessageNuGetCycleDetected" xml:space="preserve">
+    <value>NuGet restore detected a cycle. NuGet Error NU1108</value>
```
Can we provide the user more details, or explain the impact of a cycle? We should help the user understand why a cycle in NuGet Restore is bad.

ocallesp replied:
I changed this to something more verbose.

_A circular dependency A -> B -> A was detected during NuGet Restore which caused NuGet to stop restoring.
Because the NuGet restore didn't finish correctly, the project might be in a bad state.

Projects might have a package that sets AssetTargetFallback, not supported for this scenario, which caused certain PackageReference to be included.

NuGet Error NU1108_

ocallesp commented:
Some of the pending code review suggestions will be fixed as part of https://github.com/dotnet/project-system/issues/8389

nkolev92 commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/VSResources.resx:
```diff
@@ -295,4 +295,12 @@ In order to debug this project, add an executable project to this solution which
+
+This might be caused because one or more projects have a package that sets AssetTargetFallback, not supported for this scenario, which caused certain PackageReference to be included.
+
+NuGet Error NU1108</value>
```
I'd recommend removing this. 

NU1108 talks about a restore cycle within a restore graph. This merely exposes a limitation in what VS can do wrt to DT builds. 

The NU1108 text doesn't really talk about any of these issues anyways.

drewnoakes replied:
@ocallesp were you able to apply @nkolev92's feedback here?

drewnoakes replied:
This was addressed in ff84ebd58605f6878decb2619c841f4a73e96e7c.
